src/Scene/Scene.cpp

```diff
@@ -63,9 +63,15 @@ void Scene::DestroyObject(GameObject* gameObject)
                        {
                            if (pGameObject == gameObject || pGameObject->GetTransform()->IsChildOf(gameObject->GetTransform()))
                            {
-                                LOG_STREAM(DEBUG) << "Destroying GameObject " << pGameObject->GetName() << std::endl;
-                                delete pGameObject;
-                                return true;
+                               if (pGameObject == gameObject && pGameObject->GetTransform()->GetParent() != nullptr)
+                               {
+                                   // If the GameObject is the one we want to destroy, we need to remove it from its parent
+                                   // Setting reCalculateLocalTQS to false will prevent unnecessary calculations
+                                   pGameObject->GetTransform()->SetParent(nullptr, false);
+                               }
+                               LOG_STREAM(DEBUG) << "Destroying GameObject " << pGameObject->GetName() << std::endl;
+                               delete pGameObject;
+                               return true;
```

Comment:
This inner `if` can be moved out to the outer scope leading the original outer scope `if` without adding any additional comparison operations. We will need to check if this is the target object first anyway in any case. And we can use boolean alias for these criteria like:

* `found`
* `isChild`/`isChildOf`
* `nonRoot`/`hasParent`

and thus we can go like:

```
if (found && hasParent):
    removeFromParent()

if (found || isChildOf):
    delete
    return true

return false
```

Reply:
You are right.

When I was changing the code I came up with a question:

When it's iterating on child objects, if the object's parent has already been deleted, I would expect all functions using the parent's fields/functions, like `removeFromParent()` `IsChildOf()` will require a null check / throw exceptions.
However it did not and everything works without any bugs???

Reply:
In that case, I would suggest either changing the way the code works or figuring it out and writing tests for it, instead of leaving a magically working code.

For example, deferring deleting the target object after all children are gone. Or use DFS-like recursive deletion instead of BFS-like.